### DISCUSSION TITLE
Support for images

### DISCUSSION THREAD

**Discussion by EXCIDIVM** [samihsoylu/journal#27] [Ideas]
First off: Fantastic work! I actually like your version of a self-hosted journal a lot.

Just one feature is missing for me to actually use it as my daily driver: Support for images. I am very much used to add images to my journal entries. I also like having the first image to be visible in the article overview.

**Comment by samihsoylu** [comment]
Hey @EXCIDIVM, I'm glad to hear you enjoy using Journal.

Since each entry must be encrypted, adding support for images is quite tricky. 

Generally, what most platforms do is upload the photo to a storage, and reference the image in the "entry".  This means in our case, the entry is still encrypted, but the image stored in storage is not. 

Theoretically, it is possible to attempt image encryption after uploading it to a storage, but it will require some time and research and also tinkering. I am not sure it is worth the trouble.

I will think about this.

But to help you, I want to mention that, it is also possible to use an external image platform such as imgur for the storing your photos. Then you can include them in an entry like this:
```markdown
![Alt Text](url)
```
Example:
```markdown
![A Cat](http://images6.fanpop.com/image/photos/36300000/Cats-image-cats-36360540-500-750.jpg)
```

**Reply by EXCIDIVM** [reply]
Thank you for the suggested workaround!
I am not a programmer myself, but theoretically, wouldn't it be possible to add the image as binary stream inside the encrypted file? So it wouldn't need an extra file.

**Reply by samihsoylu** [reply]
Hi @EXCIDIVM 

The suggested solution would probably work if entries indeed were encrypted files on a local desktop application.

However, in practice an entry isn't an encrypted file. It is an encrypted column in a MySQL database.  **Perhaps** in theory it is possible to even store a base64 encoded binary in an entry, and have it saved to the database. But this would give little control to how many photos an entry can hold, and the [markdown](https://www.howtogeek.com/448323/what-is-markdown-and-how-do-you-use-it/) would not exactly look pretty.. 

Since journalapp.nl is open to the public, I can also expect users abusing such a feature by creating entries with large images. Ideally it would be nice to have a upload tool that takes into account the image file size too.

It is also unknown how this would effect encryption and decryption performance.

As I mentioned previously, there is a lot of work that needs to go into this, and while we are at it, it would also make sense to also upgrade the editor to a more rich based interface like the GitHub commenting system. But all this goes beyond the depth of back-end development, and I don't have the time to work on tedious features such as this one. And front-end development is not my expertise.

Perhaps someone else from the open source community is able to fork this project, to extend it and build this.

**Comment by samihsoylu** [comment]
Hi @EXCIDIVM 

This week I discovered that TinyMCE editor is in-fact very easy to integrate with the Journal platform.

An image upload feature will come in Journal v1.4 but the first image won't be visible in the entries overview

Follow it here: https://github.com/samihsoylu/journal/issues/28
